q2_demux/_demux.py

```diff
@@ -111,17 +138,74 @@ def __iter__(self):
+                    _trim_description(sequence_header.description):
                 raise ValueError(
                     'Mismatched sequence descriptions: %s and %s' %
-                    (_trim_trailing_slash(barcode_header.description),
-                     _trim_trailing_slash(sequence_header.description)))
+                    (_trim_description(barcode_header.description),
+                     _trim_description(sequence_header.description)))
 
             yield barcode_record, sequence_record
 
-    def __next__(self):
-        return next(self.barcode_generator), next(self.sequence_generator)
+
+class BarcodePairedSequenceFastqIterator(collections.abc.Iterable):
+    def __init__(self, barcode_generator, forward_generator,
+                 reverse_generator):
+        self.barcode_generator = barcode_generator
+        self.forward_generator = forward_generator
+        self.reverse_generator = reverse_generator
+
+    def __iter__(self):
+        # Adapted from q2-types
+        for barcode_record, forward_record, reverse_record \
+                in itertools.zip_longest(self.barcode_generator,
+                                         self.forward_generator,
+                                         self.reverse_generator):
+            if barcode_record is None:
+                raise ValueError('More sequences were provided than barcodes.')
+            if forward_record is None or reverse_record is None:
+                raise ValueError('More barcodes were provided than sequences.')
+            # The id or description fields may end with "/read-number", which
+            # will differ between the sequence and barcode reads. Confirm that
+            # they are identical up until the last /
+            barcode_header = _record_to_fastq_header(barcode_record)
+            forward_header = _record_to_fastq_header(forward_record)
+            reverse_header = _record_to_fastq_header(reverse_record)
+
+            # confirm that the id fields are equal
+            if not (_trim_id(barcode_header.id) ==
+                    _trim_id(forward_header.id) ==
+                    _trim_id(reverse_header.id)):
+                raise ValueError(
+                    'Mismatched sequence ids: %s, %s, and %s' %
+                    (_trim_id(barcode_header.id),
+                     _trim_id(forward_header.id),
+                     _trim_id(reverse_header.id)))
+
+            # if a description field is present, confirm that they're equal
+            if barcode_header.description is None and \
+               forward_header.description is None and \
+               reverse_header.description is None:
+                pass
+            elif barcode_header.description is None:
+                raise ValueError(
+                    'Barcode header lines do not contain description fields '
+                    'but sequence header lines do.')
+            elif (forward_header.description is None or
```

Comment:
Same comment here - can you split this into two `elif` statements so we can provide a more specific error message?